Changes for GHC 7.10

Warbo opened the pull request:
Changes include:
- Bumped the SYB dependency (no noticable effect)
- `ClassP` is missing from template-haskell 2.10, so use `AppT` instead
- Take kinded type variables into account, since derived instances in the test suite weren't including variables (e.g. `instance Serial Foo` rather than `instance (Serial a) => Serial (Foo a)`)
- Include plausible-looking `Applicative` and `Functor` instances, to obey new superclass constraints
- Annotate a test as using lists, to avoid ambiguity now that fold functions are more generic (foldable/traversible)

Note: I have _not_ tested this on other GHC versions.

Warbo committed:
Update for SYB, TH, applicative/monad and foldable/traversable

Warbo committed:
Remove dead code

jasonreich commented:
Sounds good.

Travis CI build failed because it is pinned to 7.8. Could you please update `.travis.yml` as part of your pull request.

Thanks.

Warbo committed:
Bump travis to GHC 7.10

Warbo committed:
Stop travis simplifying 7.10 to 7.1

Warbo committed:
Factor out some copypasta

Warbo committed:
Use CPP to maintain GHC 7.8 compatibility

Warbo committed:
Choose the type of applyClass depending on TH version

Warbo committed:
Reinstate GHC 7.8 fsince travis doesn't have 7.10

Warbo commented:
Finally got the build to pass. Since the changes work around different template haskell types, I've used CPP to switch between them. Checking the template-haskell version in CPP seems to cause a syntax error with GCC's preprocessor, so I've made it check GHC's version instead.

I've kept Travis testing with 7.8, since it doesn't seem to have 7.10 available. I've tested this myself on NixOS using GHC 7.10 and 7.8.4.

Warbo committed:
Clean up cabal file

No changes to dependencies or module lists, but put things on separate lines and
sort, so that diffs make more sense.

Warbo committed:
Fixes for GHC 8

Warbo committed:
New bounds for GHC 8

Warbo committed:
Benchmark fixes for GHC 8 and new Criterion

Warbo committed:
Rename .cabal file to be lowercase, matching package name

Avoids issues like https://github.com/input-output-hk/haskell.nix/issues/415

Warbo committed:
Ignore dist-newstyle too

Warbo committed:
Minor bump to 1.1.1 for .cabal file renaming

I've done this as a separate commit to the renaming, since git is bad at
noticing renames.

Warbo committed:
Rearrange TH uses, to avoid cross-chunk references

When Template Haskell is used to splice definitions into a module, that module
is no longer compiled 'all at once': instead, the part before any TH is
compiled, the part between the first two splices is compiled, and so on until
the whole module is done. The downside of splitting up the compilation like this
is that definitions in earlier 'chunks' cannot reference definitions from later
'chunks', since they won't be found during compilation. Hence why I've had to
rearrange some definitions in the suite/Functionality.hs file.

Warbo committed:
Remove CPP, as Template Haskell has changed too much

Warbo committed:
Update Template Haskell after breaking changes

Warbo committed:
Add Nix definitions

Warbo committed:
Update Nixpkgs to 23.11

Warbo committed:
Relax upper version bounds to allow newer GHC

Warbo committed:
Fix applicative/monad warnings

Warbo committed:
Fix "missing home modules" warning

Warbo committed:
Update maintainer & repo fields (this fork is now 8 years old!)

Warbo committed:
Bump version to 1.1.3, in light of fixes
